rust/chains/hyperlane-cosmos/src/providers/rpc.rs

```diff
@@ -83,58 +91,117 @@ impl CosmosWasmIndexer {
             reorg_period,
         })
     }
-}
 
-impl CosmosWasmIndexer {
-    #[instrument(level = "trace", err, skip(self))]
-    async fn tx_search(&self, query: Query, page: u32) -> ChainResult<TxSearchResponse> {
-        Ok(self
-            .provider
-            .rpc()
-            .tx_search(query, false, page, PAGINATION_LIMIT, Order::Ascending)
+    async fn get_block(client: HttpClient, block_number: u32) -> ChainResult<BlockResponse> {
+        Ok(client
+            .block(block_number)
+            .await
+            .map_err(Into::<HyperlaneCosmosError>::into)?)
+    }
+
+    async fn get_block_results(
+        client: HttpClient,
+        block_number: u32,
+    ) -> ChainResult<BlockResultsResponse> {
+        Ok(client
+            .block_results(block_number)
             .await
             .map_err(Into::<HyperlaneCosmosError>::into)?)
     }
 
+    async fn get_latest_block(client: HttpClient) -> ChainResult<BlockResponse> {
+        Ok(client
+            .latest_block()
+            .await
+            .map_err(Into::<HyperlaneCosmosError>::into)?)
+    }
+
+    // TODO: Refactor this function into a retrying provider. Once the watermark cursor is refactored, retrying should no longer
+    // be required here if the error is propagated.
+    #[instrument(err, skip(f))]
+    async fn call_with_retry<T>(
+        mut f: impl FnMut() -> Pin<Box<dyn Future<Output = ChainResult<T>> + Send>>,
+    ) -> ChainResult<T> {
+        let mut retries: u32 = 5;
+        for _ in 1..retries {
+            match f().await {
+                Ok(res) => return Ok(res),
+                Err(err) => match retries.checked_sub(1) {
+                    Some(remaining) => retries = remaining,
+                    None => break,
+                },
+            }
+            debug!("Retrying call");
```

Comment:
It's a bit annoying to return the last error, and since we'll eventually refactor this into a retrying provider I'll look into it then (added a todo). Added the suggestions to the `debug` 🫡